[Interpreter] Fall back to loading Swift dylibs from /usr/lib/swift on Apple platforms.

mikeash opened the pull request:
Continue to load the dylibs next to the compiler if they exist. If they don't, then use the OS's dylibs.

Requires LLDB PR https://github.com/apple/swift-lldb/pull/1596.

rdar://problem/47528005

mikeash committed:
[Interpreter] Fall back to loading Swift dylibs from /usr/lib/swift on Apple platforms.

Continue to load the dylibs next to the compiler if they exist. If they don't, then use the OS's dylibs.

rdar://problem/47528005

mikeash commented:
@swift-ci please test

mikeash commented:
https://github.com/apple/swift-lldb/pull/1596

@swift-ci please test

mikeash commented:
apple/swift-lldb#1596

@swift-ci please test

mikeash commented:
apple/swift-lldb#1596

@swift-ci please test

mikeash commented:
apple/swift-lldb#1596

@swift-ci please test os x platform

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -66,7 +71,11 @@ static void updateRuntimeLibraryPaths(SearchPathOptions &SearchPathOpts,
-  SearchPathOpts.RuntimeLibraryPath = LibPath.str();
+  SearchPathOpts.RuntimeLibraryPaths.clear();
+  SearchPathOpts.RuntimeLibraryPaths.push_back(LibPath.str());
+#if defined(__APPLE__) && defined(__MACH__)
```
Can you do this in terms of the target triple rather than the host architecture? I know we can only JIT in-process if they're the same, but it's the principle of the thing.

mikeash replied:
Good idea!

jrose-apple commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -1066,9 +1066,12 @@ class ParseableInterfaceModuleLoaderImpl {
-    if (resourceDir.empty()) return false;
-    return path.startswith(resourceDir);
+    for (auto &RuntimeLibraryPath : ctx.SearchPathOpts.RuntimeLibraryPaths) {
+      if (path.startswith(RuntimeLibraryPath)) {
```
This should just be using `RuntimeResourcePath` to begin with. The system libraries are *not* in the compiler resource directory.

mikeash replied:
Right you are.

jrose-apple commented on lib/Immediate/ImmediateImpl.h:
```diff
@@ -38,7 +38,7 @@ namespace immediate {
 ///
 /// \param runtimeLibPath Path to search for compiler-relative stdlib dylibs.
-void *loadSwiftRuntime(StringRef runtimeLibPath);
+void *loadSwiftRuntime(const std::vector<std::string> &runtimeLibPaths);
```
Nitpick: `ArrayRef<std::string>` would be preferred.

mikeash replied:
Good point, I'll fix it.

jrose-apple commented on lib/Immediate/Immediate.cpp:
```diff
@@ -66,8 +66,18 @@ static void *loadRuntimeLib(StringRef sharedLibName, StringRef runtimeLibPath) {
-void *swift::immediate::loadSwiftRuntime(StringRef runtimeLibPath) {
-  return loadRuntimeLib("libswiftCore" LTDL_SHLIB_EXT, runtimeLibPath);
+static void *loadRuntimeLib(StringRef sharedLibName,
+                            const std::vector<std::string> runtimeLibPaths) {
```
…which would have avoided the accidental copy here.

mikeash committed:
[Interpreter] Clean up the /usr/lib/swift fallback code.

- Check the target triple at runtime to decide whether to use the fallback.
- Change isInResourceDir to actually check the resource dir.
- Use ArrayRef<std::string> instead of std::vector<std::string>.

mikeash commented:
apple/swift-lldb#1596

@swift-ci please test

mikeash commented:
@jrose-apple How's that look?

jrose-apple reviewed (approved):
Thanks!

mikeash commented:
apple/swift-lldb#1596

@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
Forgot to mention the LLDB PR this time!

mikeash commented:
https://github.com/apple/swift-lldb/pull/1596

@swift-ci please test
